common/locks/priority_semaphore_impl.go

```diff
@@ -0,0 +1,136 @@
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package locks
+
+import (
+	"context"
+	"sync"
+)
+
+type Priority int
+
+const (
+	PriorityHigh Priority = iota
+	PriorityLow
+)
+
+type PrioritySemaphoreImpl struct {
+	capacity    int
+	highWaiting int         // Current number of high priority requests waiting.
+	highCount   int         // Current count of high-priority holds
+	lowCount    int         // Current count of low-priority holds
+	lock        sync.Locker // Mutex to protect conditional variable and changes to the counts
+	highCV      *sync.Cond  // Condition variable for high-priority tasks
+	lowCV       *sync.Cond  // Condition variable for low-priority tasks
+}
+
+var _ PrioritySemaphore = (*PrioritySemaphoreImpl)(nil)
+
+// NewPrioritySemaphore creates a new PrioritySemaphoreImpl with specified capacity
+func NewPrioritySemaphore(capacity int) PrioritySemaphore {
+	mu := &sync.Mutex{}
+	highCV := sync.NewCond(mu)
+	lowCV := sync.NewCond(mu)
+	return &PrioritySemaphoreImpl{
+		capacity:  capacity,
+		highCount: 0,
+		lowCount:  0,
+		lock:      mu,
+		highCV:    highCV,
+		lowCV:     lowCV,
+	}
+}
+
+func (s *PrioritySemaphoreImpl) Acquire(ctx context.Context, priority Priority, n int) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	for s.lowCount+s.highCount+n > s.capacity {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			if priority == PriorityHigh {
+				s.highWaiting++
+				s.highCV.Wait()
+				s.highWaiting--
+			} else {
+				s.lowCV.Wait()
+			}
+		}
+	}
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+	if priority == PriorityHigh {
+		s.highCount += n
+	} else {
+		s.lowCount += n
+	}
+	return nil
+}
+
+func (s *PrioritySemaphoreImpl) TryAcquire(priority Priority, n int) bool {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.lowCount+s.highCount+n > s.capacity {
+		return false
+	}
+
+	if priority == PriorityHigh {
+		s.highCount += n
+	} else {
+		s.lowCount += n
+	}
+	return true
+}
+
+func (s *PrioritySemaphoreImpl) Release(priority Priority, n int) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if priority == PriorityHigh {
+		if n > s.highCount {
+			panic("trying to release more resources than acquired")
+		}
+		s.highCount -= n
+
+	} else {
+		if n > s.lowCount {
+			panic("trying to release more resources than acquired")
+		}
+		s.lowCount -= n
+	}
+
+	if s.highWaiting != 0 {
+		// If n > 1, more than one caller could potentially make progress.
+		for ; n > 0; n-- {
+			s.highCV.Signal()
+		}
+	} else {
+		for ; n > 0; n-- {
+			s.lowCV.Signal()
+		}
+	}
```

Comment:
don't you want to signal up to highWaiting on highCV and then signal up to lowWaiting rest on lowCV? this looks like it'll only signal high if there's enough room for some high and some low

Reply:
If n > highWaiting then we could let both high and low waiters through too, which this implementation can't do

Reply:
We don't want to singal upto highWaiting threads. If this function call only releases 2 resources, we just need to wake 2 highWaiting threads. If there is no one waiting on highCV, then it will wake up threads in lowCV.

n in this case is the number of resources getting released in this Release() call.

Reply:
right, but what if you release 2, and there's one highWaiting and one lowWaiting? you should signal both, no?

Reply:
Yes, exactly

Reply:
Yeah got it! Made the change.
I was thinking only about the case n=1 when I wrote this condition. Then forgot to consider this case.